### PR TITLE
compact: Add concurrency to GatherNoCompactFilter

### PR DESCRIPTION
**Pull request by zecke**
Follow the pattern used for IgnoreDeletionMarkFilter but instead
of stopping early completely drain the channel and report any of
the errors that occurred.

* [ ] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Verification

Synced metadata in prod (but we didn't have no compact marked blocks)

### CODE REVIEW

**Review by pracucci — APPROVED** [review]
LGTM, thanks!

**Review by bwplotka — APPROVED** [review]
Thanks, nice, high-quality code!

Not sure if this is necessary, there is some extra allocations and complexity here. Happy to merge, if other maintainers are happy with it.

Ideally for such work, we have quick benchmarks too, next time (:

**Review by kakkoyun — APPROVED** [review]

**Comment by kakkoyun** [comment]
Hey @zecke. Is there a chance to add some benchmarks to this before we merge?

**Comment by zecke** [comment]
I can look at a benchmark but not this week.

**Comment by bwplotka** [comment]
Let's think about it later on (:

**Pull request merged**
